### ISSUE TITLE
Control charts/Shewhart charts

### ISSUE THREAD

**Issue by TMiguelT** [ewels/MegaQC#78]
I've been reading a few papers on QC, and it occurred to me that the main thing missing from MegaQC is actually the absence of a clear "control chart" (pictured). I have a feeling that this is actually the first thing a user might expect from a tool like MegaQC, but it seems to be mysteriously missing. 

![image](https://user-images.githubusercontent.com/5019367/57922790-e7993e80-78e3-11e9-85e0-8d9da9e24530.png)

We currently have the framework for this, in the form of the "trend" plots, but the standard deviation and mean aren't present. If this were added, it would also allow a rudimentary implementation of Westgard rules, by interrogating the plot, which is why I think this is the place to start. My question is: should these features be added to the existing trend plots, or should "control chart" be added as a new core MegaQC mode?

**Comment by ewels** [comment]
This has always been planned for the trend plots (it was in fact one of the key reasons for adding trend plots), we just never got to writing it.

My original idea was to have basically exactly what you've drawn, with dropdowns to select different options for each line. eg: mean, median, manual / SD, SE, _n_ th percentile, manual etc. In my mind it's basically the same feature as that described in #28 - I was imaging people setting these thresholds visually and then clicking to create an automated alert based on them.

> _it seems to be mysteriously missing._

This is why MegaQC doesn't have a first release yet - much of the original implementation plan has never been completed due to staff changes. I still hope to get there one day!

**Comment by TMiguelT** [comment]
Okay, so lets say my thresholds feature supports aggregate operations like mean, median, SD. Is this as simple as selecting from the database all thresholds that use this data type, and then annotating it onto the plot? If there are no thresholds, should we default to a mean + SD?

**Comment by ewels** [comment]
I was more thinking of just using the plots to create and save the thresholds - more of a one-way road. Though could be tied in to the favourite plot mechanism. That has a page to list favourite plots so they could be added there as a subsection I think.

Don't want to limit to a single set of thresholds for a given plot and don't want to auto-add these in the main interface. Should be static once saved. Does that make sense? I can see it in my head but not sure that I'm describing it very well...

**Comment by TMiguelT** [comment]
I know that using plots to create the threshold would look visually neat, but I suspect that most users will have a specific number or statistic in mind (standard deviations etc) that they will want to apply to the plot, rather than going by what looks right.

This is why I was suggesting that we just apply my threshold/alert objects to each plot they apply to, in the form of a line of some sort. This would also make them static since the thresholds themselves are stored in the db. The only thing I disagree with is auto-adding thresholds. I think by default 1 standard deviation would be a nice thing to see.

**Comment by ewels** [comment]
I think we're agreeing with one-another here. I mean to enter numbers / statistics manually, just in combination with the plot (it will usually want to be combined with a specific data type and filtering set, so seeing this in a plot is good validation before saving).

> This is why I was suggesting that we just apply my threshold/alert objects to each plot they were designed for to

I guess I see almost every plot as being unique, hence my workflow above.

But yeah, I think we're on the same page 👍

**Comment by TMiguelT** [comment]
So you think that in the data model, thresholds should be attached to the favourite plots? The reason I don't like this is because thresholds won't *just* be used to provide upper and lower limits on the plots, they will also be used to generate automated alerts, and these alerts aren't tied to favourite plots. Do you see my dilemma? 

I think that thresholds should be an independent entity that is auto-drawn on plots that use that data type. Can you think of a better option that also supports automated alerts?

**Comment by ewels** [comment]
> Do you see my dilemma?

Not really 😅 

For the thresholds, you need the following metadata:

* Plot type (trend / distribution / etc)
* Data fields
* Sample filtering
* Thresholds
* Actions

The first three _are_ Favourite Plots. These aren't static, for example they are used to build dashboards with live data etc.

> I think that thresholds should be an independent entity that is auto-drawn on plots that use that data type.

I don't really understand how you can _auto-draw on data of that type_. You'd have to go to the relevant plot page (trend plot etc), then select the correct data fields, then select the exact sample filtering, and then the thresholds would appear? This seems very clunky to me..? Instead, using the existing code for Favourite Plots we could just add on the final two points.

Behind the scenes, the thresholds could be separate from the favourite plots, or probably just refer to a given favourite plot in order to collect the first three metadata fields.

**Comment by TMiguelT** [comment]
> I don't really understand how you can auto-draw on data of that type. You'd have to go to the relevant plot page (trend plot etc), then select the correct data fields, then select the exact sample filtering, and then the thresholds would appear? 

We'd have a Threshold model that has an associated data type (I think it's called `SampleDataType` in the db) and a value (defined either as an absolute threshold or a statistical one). This threshold doesn't actually need to store a specific sample filtering, or a *list* of other fields shown on the plot, because the threshold is a universal rule associated with the data *type*, not with this specific plot. 

I don't really understand why you think thresholds should be associated with the plots themselves. Think of the thresholds as laboratory SOPs (Standard Operating Procedure). They can't just be applied in the context of one plot, they have to be a universal rule, e.g. that all genes have a mean coverage of X. This is why they need to be an independent concept. I can see a situation where you'd want to have different SOPs on a per-test basis (e.g. a different sequencing kit that covers different genes), so that could be added to the threshold entity. In that situation, where you have a complex set of filters, it might be hard to autodetect which thresholds to show. But in this case maybe we just provide a list of thresholds that might be relevant (because they apply to a `SampleDataType` you're showing), and let the user turn them on and off?

**Comment by TMiguelT** [comment]
Actually, I'm wondering if control charts should be considered a different plot type to a trend plot. The reason being, in a control chart you expect (and hope) that the mean stays the same, and hence you want to plot the mean and stdev over the whole dataset. This means that the mean will be a flat horizontal line: 
![image](https://user-images.githubusercontent.com/5019367/58403938-43fc1b00-80a7-11e9-9f60-72a429912b68.png)

However in a trend plot, you're actively looking for trends, meaning you actually want a *trendline*, not a mean, which might be moving up or down:
![image](https://user-images.githubusercontent.com/5019367/58404038-87568980-80a7-11e9-87cc-dad11dcc1a78.png)

What do you think?

**Comment by TMiguelT** [comment]
My current thoughts on the alert thresholds are that we should be able to pull out the correct information from the filter, in order to annotate it onto a plot.

For example, say the lab has an SOP which is "all samples for this kidney disease test should have a minimum of 30X gene coverage". In the query language, this translates to something like (I'm mostly making this up):
```json
[
  [
    {
      "type": "samplemeta",
      "key": "custom__test type",
      "cmp": "eq",
      "value": "kidney"
    },
    {
      "type": "samplemeta",
      "key": "samtools__average_coverage",
      "cmp": "lt",
      "value": 30
    }
  ]
]
```

So, from this information, we can run the filter on any incoming or existing data to find samples that fail the SOP, and create an alert.

In addition, we can automatically parse this filter array, and if the user makes a plot that has the same `test_type == "kidney"` filter, and is plotting the `samtools__average_coverage` field, then we add this filter as a threshold to that plot.

There could well be the option to enable/disable thresholds manually, beyond this automatic annotation, and attach that information to a favourite chart, to better customize it. But that manual aspect isn't actually required, in my opinion.

Thoughts?

**Comment by ewels** [comment]
Yeah, sounds good! I think it's probably a good idea to make the filters and alerts use as much of the same machinery as possible (avoid code duplication + easier maintenance). But as you say, they should be almost the same thing really.

**Comment by TMiguelT** [comment]
Yeah, so I'm suggesting that the filter used by AlertThreshold is the same JSON format used by sample filters and favourite plots.

**Comment by ewels** [comment]
Nice! @tsnowlan - if you have any thoughts or have time to go over the open PRs, that would be very much appreciated 👍

**Comment by TMiguelT** [comment]
@ewels the main thing I'm interested in is a preference of which user interface technology you would prefer me to use going forward. I've prototyped using React (#83) and Dash (#81). Both have upsides and downsides in terms of maintainability and verbosity. However since you are heading the development I think it should be something you consider yourself, and decide if either technology is something you think you can build on going forward. If you don't think either is reasonable then that's fine too.

If you're already familiar with React and Dash then I don't think reviewing them will take more than 10 minutes each. And it would really help me out.

**Comment by ewels** [comment]
Yup, got it. Will take more than 10 minutes though as I need to set everything up on my home computer and generate example test data / get the base version of MegaQC working first before I can do any testing. Plus it has to compete with other critical work tasks at a time when I’m not meant to be doing any work ;)

**Comment by TMiguelT** [comment]
I can share some of my auto-generated multiQC files (from fastQC data) if you would find that helpful

**Comment by TMiguelT** [comment]
Also there's a chance you don't really need to run the code - the final result looks the same in both cases. It's more about the maintainability of the code and associated frameworks.

**Comment by TMiguelT** [comment]
Closed (on an initial level) by #83
